Make restart_dosbox() go through the normal shutdown routine

weirddan455 opened the pull request:
# Description

Restarting either by `config -r` or by Ctrl+Alt+Home hotkey now goes through the normal shutdown procedure. A child process gets created that waits for the parent process to terminate before continuing.

What caused me to look into this was the command line history not being saved on restart but I was concerned about other clean-up routines not being run (ex. some of the file system meta-data doesn't get written until close).

This ensures that restart is the same as a normal shutdown. It also delays init of the new process until the old process is fully terminated.

## Related issues

Fixes #4318


# Release notes

Restarting DOSBox now correctly saves the command line history.

# Manual testing

Command line history is now saved after restart on Windows and Linux. Log output confirms the parent process gets fully shutdown before the child process starts init.

The change has been manually tested on:

- [x] Windows
- [x] macOS
- [x] Linux


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] my change has been manually tested on Windows, macOS, and Linux.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

johnnovak commented:
> It's hipster Unix and they're too cool to have a `/proc` apparently.

Erm, excuse me? 😂 

https://www.opengroup.org/openbrand/register/brand3710.htm

It's _more_ Unix that Linux mate 😎🤘🏻 Well, at least it's certified (whatever that means, lol 😂)

johnnovak commented:
@weirddan455 We could always use the filesystem on macOS: generate a "pid" file in `/tmp` then poll that file in the child process until it gets deleted

weirddan455 commented:
> @weirddan455 We could always use the filesystem on macOS: generate a "pid" file in `/tmp`, delete the file on exit, then poll that file in the child process until it gets deleted.

But when is "on exit"? The last bit of our code that runs from what I can tell is `atexit(QuitSDL);` but then global destructors still run after that.

If I had a defined exit point in our code, I would just defer `execvp()` until then rather than making a temp file but it's more complicated than that.

johnnovak commented:
@weirddan455 Magic AI generated answer how to do this on a BSD (macOS is a BSD, effectively). @kklobe is now proud of me, I'm sure 😛 

Maybe it's all humbug, but hey, took me 5 seconds 😆 

<img width="711" alt="image" src="https://github.com/user-attachments/assets/f3c1a4ea-4f0a-4025-be7f-e4a90cbe72e1" />

<img width="668" alt="image" src="https://github.com/user-attachments/assets/c19a01ea-842d-4e3e-8019-53f973564da1" />

<img width="663" alt="image" src="https://github.com/user-attachments/assets/5c657e7c-1b2e-41e9-9ec7-97a535388269" />

weirddan455 commented:
Ready for review. Updated toplevel comment, Added and tested Windows codepath. Mac will take the same codepath as Linux but I need a Mac user to test.

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -4774,6 +4774,33 @@ int sdl_main(int argc, char* argv[])
+		while (1) {
+			// Signal of 0 does not actually send a signal.
+			// It only checks for existance and permissions of the PID.
+			constexpr int SIGNAL = 0;
```
Signal

weirddan455 replied:
fixed

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -4774,6 +4774,33 @@ int sdl_main(int argc, char* argv[])
+			CloseHandle(process);
+		}
+		#else
+		while (1) {
```
No love for `for (;;)`? 😎

weirddan455 replied:
I don't really have a preference. I changed it for you :smile:

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -4570,11 +4570,38 @@ static int edit_primary_config()
 
 	control->ApplyQueuedValuesToCli(parameters);
 
+	// Remove any existing --waitpid paramaters.
```
typo: parameters

weirddan455 replied:
fixed

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -4646,13 +4673,24 @@ void restart_dosbox(std::vector<std::string> &parameters) {
+			// Newly created child process immediately executes a new instance of DOSBox.
+			// It is not safe for a child process in a multi-threaded program to do much else.
+			if (execvp(newargs[0], newargs) == -1) {
+				E_Exit("Restart failed: execvp failed: %s", strerror(errno));
```
Any reason for the hard exit? Wouldn't `LOG_ERR` be better?

weirddan455 replied:
This is in the child process. It crashes if you try to do a normal exit. The parent process will do a normal exit regardless if we've gotten this far. If the child fails `execvp()`, we have to kill it.

weirddan455 commented on src/gui/sdlmain.cpp:
```diff
@@ -4570,11 +4570,38 @@ static int edit_primary_config()
+			++it;
+		}
+	}
+}
```
I had looked at that before I wrote this. It's tied to the CommandLine class and doesn't do exactly what I need anyway.

johnnovak replied:
Fair enough.

FeralChild64 commented on src/gui/sdlmain.cpp:
```diff
@@ -4570,11 +4570,38 @@ static int edit_primary_config()
+			it = parameters.erase(it);
+			if (it != parameters.end()) {
+				auto &pid = *it;
+				if (!pid.empty() && pid[0] >= '0' && pid[0] <= '9') {
```
`std::isdigit(pid[0])`?

weirddan455 replied:
Sure. Changed.

weirddan455 commented on src/gui/sdlmain.cpp:
```diff
@@ -4774,6 +4774,33 @@ int sdl_main(int argc, char* argv[])
+			}
+			Delay(50);
+		}
+		#endif
```
Formatted

johnnovak commented:
Is there anything outstanding on this @weirddan455 ?

weirddan455 commented:
> Is there anything outstanding on this @weirddan455 ? I've confirmed it works on macOS as expected.

Nope, this just slipped my mind. Merged.
